File: solutions/Leetcode_150/leetcode_150.py
from typing import List


class Solution1:
    def eval_RPN(self, tokens: List[str]) -> int:
        numbers = []
        for token in tokens:
            if token in ('+', '-', '*', '/'):
                num2 = numbers.pop()
                num1 = numbers.pop()
                if token == '+':
                    numbers.append(num1 + num2)
                elif token == '-':
                    numbers.append(num1 - num2)
                elif token == '*':
                    numbers.append(num1 * num2)
                else:
                    numbers.append(int(num1 / num2))
            else:
                numbers.append(int(token))
        return numbers.pop()


if __name__ == '__main__':
    tokens = ["10", "6", "9", "3", "+", "-11",
              "*", "/", "*", "17", "+", "5", "+"]
    ans = Solution1().eval_RPN(tokens)
    print(ans)

File: solutions/Leetcode_150/leetcode_150.cc
#include <functional>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>


using namespace std;


class Solution1
{
public:
    int evalRPN(vector<string> &tokens)
    {
        unordered_map<string, function<int (int, int)>> mapping = {
            {"+", [](int a, int b){return a + b;}},
            {"-", [](int a, int b){return a - b;}},
            {"*", [](int a, int b){return a * b;}},
            {"/", [](int a, int b){return a / b;}},
        };

        stack<int> numbers;
        for (auto token : tokens)
        {
            if (!mapping.count(token))
            {
                numbers.push(stoi(token));
            } else {
                int right = numbers.top();
                numbers.pop();
                int left = numbers.top();
                numbers.pop();
                numbers.push(mapping[token](left, right));
            }
        }
        return numbers.top();
    }
};